libp2p/io/msgio.py

```diff
@@ -1,24 +1,102 @@
-from libp2p.network.connection.raw_connection_interface import IRawConnection
+"""
+``msgio`` is an implementation of `https://github.com/libp2p/go-msgio`.
 
-from .exceptions import MissingLengthException, MissingMessageException
+from that repo: "a simple package to r/w length-delimited slices."
+
+NOTE: currently missing the capability to indicate lengths by "varint" method.
+"""
+# TODO unify w/ https://github.com/libp2p/py-libp2p/blob/1aed52856f56a4b791696bbcbac31b5f9c2e88c9/libp2p/utils.py#L85-L99  # noqa: E501
+from typing import Optional, cast
+
+from libp2p.io.abc import (
+    Closer,
+    ReadCloser,
+    Reader,
+    ReadWriteCloser,
+    WriteCloser,
+    Writer,
+)
+from libp2p.io.utils import read_exactly
 
 SIZE_LEN_BYTES = 4
+BYTE_ORDER = "big"
 
-# TODO unify w/ https://github.com/libp2p/py-libp2p/blob/1aed52856f56a4b791696bbcbac31b5f9c2e88c9/libp2p/utils.py#L85-L99  # noqa: E501
+
+async def read_length(reader: Reader) -> int:
+    length_bytes = await read_exactly(reader, SIZE_LEN_BYTES)
+    return int.from_bytes(length_bytes, byteorder=BYTE_ORDER)
 
 
-def encode(msg_bytes: bytes) -> bytes:
+def encode_msg_with_length(msg_bytes: bytes) -> bytes:
     len_prefix = len(msg_bytes).to_bytes(SIZE_LEN_BYTES, "big")
     return len_prefix + msg_bytes
 
 
-async def read_next_message(reader: IRawConnection) -> bytes:
-    len_bytes = await reader.read(SIZE_LEN_BYTES)
-    if len(len_bytes) != SIZE_LEN_BYTES:
-        raise MissingLengthException()
-    len_int = int.from_bytes(len_bytes, "big")
-    next_msg = await reader.read(len_int)
-    if len(next_msg) != len_int:
-        # TODO makes sense to keep reading until this condition is true?
-        raise MissingMessageException()
-    return next_msg
+# NOTE: temporary for this PR
+encode = encode_msg_with_length
+
+
+# NOTE: temporary for this PR
+async def read_next_message(reader: Reader) -> bytes:
+    length = await read_length(reader)
+    return await reader.read(length)
+
+
+class MsgIOWriter(Writer, Closer):
+    write_closer: WriteCloser
+
+    def __init__(self, write_closer: WriteCloser) -> None:
+        super().__init__()
+        self.write_closer = write_closer
+
+    async def write(self, data: bytes) -> int:
+        await self.write_msg(data)
+        return len(data)
+
+    async def write_msg(self, msg: bytes) -> None:
+        data = encode_msg_with_length(msg)
+        await self.write_closer.write(data)
+
+    async def close(self) -> None:
+        await self.write_closer.close()
+
+
+class MsgIOReader(Reader, Closer):
+    read_closer: ReadCloser
+    next_length: Optional[int]
+
+    def __init__(self, read_closer: ReadCloser) -> None:
+        # NOTE: the following line is required to satisfy the
+        # multiple inheritance but `mypy` does not like it...
+        super().__init__(read_closer)  # type: ignore
```

Comment:
You could do this the gross way and do `ParentClass.__init__(self, read_closer)`

(though I can't figure out which parent class `__init__` this is intended to call since neither `Reader` or `Closer` seem to implement `__init__`

Reply:
yeah, this part is weird which is why i'm not certain `mypy` will deal w/ it properly...

given the multiple inheritance in `MsgIOReadWriter`, the call chain is:

```
MsgIOReadWriter.__init__ -> MsgIOReader.__init__ -> MsgIOWriter.__init__
```

and `mypy` doesn't like that if you just statically look at line 71, the super class is actually `object` (which has a niladic `__init__`)....

i'm open to changing this and want to think about it some more...